src/rules.ml

```diff
@@ -0,0 +1,46 @@
+open! Stdune
+
+module Dir_rules = struct
+  type t = unit -> unit
+
+  let empty = (fun () -> ())
+  let union f g () = f (); g ()
+end
+
+module T = struct
+  type t = Dir_rules.t Path.Build.Map.t
+
+  let empty = Path.Build.Map.empty
+
+  let union_map a b ~f =
+    Path.Build.Map.union a b ~f:(fun _key a b -> Some (f a b))
+
+  let union =
+    union_map ~f:(fun rule1 rule2 -> fun () -> rule1 (); rule2 ())
```

Comment:
How about using `Dir_rules.union` here? Perhaps it would be nicer to make `Dir_rules.t` abstract btw.

Reply:
Yeah, `Dir_rules.union` didn't exist when this line was written.

This is getting rewritten in #2089 anyway so I wouldn't worry too much.